lib/taskana-core/src/main/java/pro/taskana/impl/ClassificationServiceImpl.java

```diff
@@ -351,4 +352,78 @@ public Classification getClassificationByTask(Task task) throws ClassificationNo
         }
     }
 
+    @Override
+    public void deleteClassification(String classificationKey, String... domains) throws ClassificationInUseException, NotAuthorizedException, ClassificationNotFoundException {
+        if (domains.length == 0) {
+            throw new IllegalArgumentException("You have to specify a domain");
+        }
+        for (String domain : domains) {
+            if (domain.equals("")) {
+                // Alternative Solution: open "deleteRootClassification".
+                // But in my opinion there is a big difference between deleting a specification of a Classification and deleting the whole classification.
+                throw new IllegalArgumentException("You cannot delete the Root Classification this way. Please use 'deleteRootClassification' instead");
+            }
+        }
+
+        TaskServiceImpl taskService = (TaskServiceImpl) taskanaEngineImpl.getTaskService();
+        boolean existingTasksForClassificationInDomains = ((TaskServiceImpl) taskService).isThereATaskForThisClassificationKeyInTheseDomains(classificationKey, domains);
+        if (existingTasksForClassificationInDomains) {
+            throw new ClassificationInUseException("There are Tasks which belong to this classification. Please complete them and try again.");
+        }
+
+        this.internalDeleteClassification(classificationKey, domains);
+    }
+
+    @Override
+    public void deleteRootClassification(String classificationKey) throws ClassificationInUseException, NotAuthorizedException {
+        String[] domains = classificationMapper.getDomainsForClassificationKey(classificationKey, CURRENT_CLASSIFICATIONS_VALID_UNTIL);
+
+        TaskServiceImpl taskService = (TaskServiceImpl) taskanaEngineImpl.getTaskService();
+        List<Task> tasks = new ArrayList<>();
+        tasks.addAll(taskService.createTaskQuery().classificationKeyIn(classificationKey).list());
+        if (!tasks.isEmpty()) {
+            throw new ClassificationInUseException("There are Tasks which belong to this classification. Please complete them and try again.");
+        }
+
+        String newParent = classificationMapper.findByKeyAndDomain(classificationKey, "", CURRENT_CLASSIFICATIONS_VALID_UNTIL).getParentClassificationKey();
+
+        try {
+            this.internalDeleteClassification(classificationKey, domains);
+        } catch (ClassificationNotFoundException e) {
+            // should not happen.
+            throw new InternalError("A internal error occured. There may be a problem with the 'getDomainsForClassificationKey' request of the classificationMapper.");
+        }
+
+        this.careForTheChildren(classificationKey, newParent);
```

Comment:
Is a calling this function in every case (Exception) really necessary? Correct me if i´m wrong.

Reply:
Why don't we use a recursive function here?